### PR TITLE
[Material] Add Material WB

### PR DESCRIPTION
**Pull request by davesrocketshop**
**NOTE**: This PR introduces a new dependency on _yaml-cpp_ (https://github.com/jbeder/yaml-cpp). This is available through
 apt on Debian based distributions, but the LibPack image will need to be updated for Windows builds

First PR for Issue #10174 - the Materials system update.

- The editor is minimally functional. You can save changes but may need to restart for them to take effect
- The API supports new and old methods of access. See the unit tests in src/Mod/Material/TestMaterialApp.py for examples
- Uses an updated .FCMat format. The old editor won't read the new files, but the new editor will still read the old files.
- Defaults to the old editor. This can be changed in the Material Preferences pane
- System materials are in the new format. You can still use the old format files by copying them to a separate folder and adding that as a custom folder in the Material Preferences pan
- Moves the preferences from FEM to a separate Materials section. This was submitted as a PR prior to 0.21 but that PR has been closed to avoid conflicts

### CODE REVIEW

**Comment by chennes** [comment]
With the new dependency it's going to take a day or two before we can get the CI working with it. Is yaml-cpp v0.8.0 OK?

**Comment by davesrocketshop** [comment]
I expected there to be some delay until the dependencies are met. I've been testing with v0.7.0, so v0.8.0 should be fine. I'll update mine and test to verify.

**Comment by davesrocketshop** [comment]
Tested with v0.8.0

This version corrected some build issues I had to kludge to make work with 0.7.0, so this is good.

**Comment by chennes** [comment]
Well, not really so good, we have to support Ubuntu 20.04 for couple of years yet, and it only provides yampl-cpp 0.6. Ubuntu 22.04, 22.10, and 23.04 support yaml-cpp 0.7, and no one has an apt package for 0.8 yet. So although I can easily at 0.8 to the Windows LibPack, for those working on Linux and using apt, we'll have to support both 0.6 and 0.7 for some time to come. I have no idea whatsoever this is going to entail for this Mod, however, I have not looked at the code.

**Comment by davesrocketshop** [comment]
I have been building successfully on Ubuntu 20.04, but due to server failures I haven't been able to test as much as previously. It may not be much of a problem however the concern is noted.

**Comment by chennes** [comment]
See https://github.com/FreeCAD/FreeCAD/pull/10380 for Linux CI. Windows will take a bit longer.

File: src/Mod/Fem/Gui/AppFemGui.cpp [hunk 1]
```diff
@@ -176,8 +175,7 @@ PyMOD_INIT_FUNC(FemGui)
     new Gui::PrefPageProducer<FemGui::DlgSettingsFemElmerImp>(QT_TRANSLATE_NOOP("QObject", "FEM"));
     new Gui::PrefPageProducer<FemGui::DlgSettingsFemMystranImp>(QT_TRANSLATE_NOOP("QObject", "FEM"));
     new Gui::PrefPageProducer<FemGui::DlgSettingsFemZ88Imp>(QT_TRANSLATE_NOOP("QObject", "FEM"));
-    new Gui::PrefPageProducer<FemGui::DlgSettingsFemMaterialImp>(QT_TRANSLATE_NOOP("QObject", "FEM"));
-
+
```
Comment (luzpaz) [inline src/Mod/Fem/Gui/AppFemGui.cpp:178]: Trailing whitespace

Comment (luzpaz) [reply src/Mod/Fem/Gui/AppFemGui.cpp:178]: See the rest of the trailing whitespaces in https://github.com/FreeCAD/FreeCAD/actions/runs/5983698946/job/16234124245?pr=10368#step:5:1

**Comment by chennes** [comment]
I've built a new LibPack... now I have to integrate into our CI system, then re-run this PRs CI. Hopefully that all goes smoothly and your checkmarks will all be green soon.

**Comment by adrianinsaval** [comment]
please notify me when this PR is merged so that I can add yaml-cpp to the conda and flatpak builds.

**Comment by chennes** [comment]
@0penBrain I need some help with this one -- I've rolled a new LibPack that includes yaml-cpp and fmt, and it looks to me like the CI is downloading it, but then it's not finding either of those libraries. Is it somehow using a cached version of the old libpack? I don't understand the caching as it is set up here.

**Comment by 0penBrain** [comment]
> I've rolled a new LibPack that includes yaml-cpp and fmt, and it looks to me like the CI is downloading it, but then it's not finding either of those libraries. Is it somehow using a cached version of the old libpack?

Yes, it uses a cached version. Process to update cache is documented here :smile: : https://github.com/FreeCAD/FreeCAD/blob/ea116dc0332ccfc5797f248b04492a16f421b787/.github/workflows/actions/windows/getLibpack/action.yml#L28-L30

Basically, you go to https://github.com/FreeCAD/FreeCAD/actions/caches and delete the libpack cache so it is updated next time:
image

I will do it now and re-run this PR check.

**Comment by chennes** [comment]
Thanks (and sorry I apparently can't read 😒) -- Windows build is passing now. Last it seems
I have something to update in the linter CI, then we should be ready to merge.

**Comment by chennes** [comment]
@davesrocketshop I haven't checked it -- are you using our .clang-format here? If so then after merging I'll get it added to our pre-commit.

**Comment by davesrocketshop** [comment]
> @davesrocketshop I haven't checked it -- are you using our .clang-format here? If so then after merging I'll get it added to our pre-commit.

I haven't been but I will going forward (now that I know it's there)

**Comment by chennes** [comment]
In the FCMat files, how would you feel about splitting up the Author and License lines, and using an [SPDX identifier](https://spdx.org/licenses/) for the licenses?

**Comment by davesrocketshop** [comment]
> In the FCMat files, how would you feel about splitting up the Author and License lines, and using an [SPDX identifier](https://spdx.org/licenses/) for the licenses?

That would be an easy change. The only issue would be automatic conversion of existing files, and there is a potential although low risk of backwards compatibility. Realistically that shouldn't be a big deal though.

**Comment by chennes** [comment]
One of the reasons for it is that some distros are quite strict about what licenses they are willing to distribute, and it would be nice to make it easy to identify the things they have to remove. A separate field and standardized tag makes that a straightforward exercise for them. I'm doing some work in this direction with the Addon Manager as well.

**Comment by davesrocketshop** [comment]
What support is in FC currently, such as a list of identifiers? Have time for a Zoom to discuss? (Eastern time zone)

**Comment by chennes** [comment]
The main body of the code has this License.h file, used for a license selector for the document as a whole:
https://github.com/FreeCAD/FreeCAD/blob/master/src/App/License.h

The Addon Manager Developer Mode has a license selector with a smaller selection:
https://github.com/FreeCAD/FreeCAD/blob/master/src/Mod/AddonManager/addonmanager_devmode_license_selector.py

I can set up a Zoom if you'd like (US Central timezone) -- drop me a line at my username at freecad.org

**Comment by davesrocketshop** [comment]
Something screwy is happening in the unit tests that I'm not able to reproduce. One error that passes on all architectures for me:

 ```
ERROR: testModelLoad (TestMaterialsApp.MaterialTestCases)
 ----------------------------------------------------------------------
 Traceback (most recent call last):
  File "/usr/local/Mod/Material/TestMaterialsApp.py", line 37, in testModelLoad
    density = self.ModelManager.getModel("454661e5-265b-4320-8e6f-fcf6223ac3af")
 LookupError: Model not found

```

So I changed it to assert that the error that shouldn't be thrown is actually thrown:

```
ERROR: testModelLoad (TestMaterialsApp.MaterialTestCases)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/Mod/Material/TestMaterialsApp.py", line 38, in testModelLoad
    self.assertRaises(LookupError, self.ModelManager.getModel("454661e5-265b-4320-8e6f-fcf6223ac3af"))
LookupError: Model not found

```
This should not be a failure! And there's no way both should fail.

I can't get the build to show any debugging information as it's all stripped from the log file.

Suggestions?

**Comment by chennes** [comment]
I'm trying to get these tests running on my system, and it's giving me some difficulty: this may or may not be related to the problem you are seeing (though I can reproduce that locally in a debug build, so I will continue to poke at it). Some observations:
* yaml-cpp is being specified incorrectly in the cmake files, it isn't picking up the debug library when it should be, it always tries to use the release version. This might be a LibPack-specific problem.
* When running TestMaterialsApp for the first time, I get the following output:
```
18:35:26  Empty filename passed to function
18:35:26  Empty filename passed to function
18:35:26  YAML parsing error: 'G:/FreeCAD/chennes/FreeCAD-chennes/src/Mod/Material/FluidMaterial/Air.FCMat'
```
I have no idea where that path is coming from, it's non-existent, and actually doesn't bear much resemblance to a real path on my system, and certainly no the installation path of this build.
* When running tests for the first time, it takes a very long time to get started: maybe ten minutes? It is spending all that time in this loop:
```
void ModelLoader::loadLibrary(const ModelLibrary& library)
{
    if (_modelEntryMap == nullptr) {
        _modelEntryMap = new std::map<QString, ModelEntry*>();
    }

    QDirIterator it(library.getDirectory(), QDirIterator::Subdirectories);
    while (it.hasNext()) {
        auto pathname = it.next();
        QFileInfo file(pathname);
        if (file.isFile()) {
```
* At the end of the processing, I get an "Invalid string position" exception
* If I run the tests a second time without restarting FreeCAD it runs much faster, but I get your failure from above:
```
19:19:47  FAILED: testCalculiXSteel (TestMaterialsApp.MaterialTestCases)
Traceback (most recent call last):
  File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 60, in testPartExecutor
    yield
  File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 633, in _callTestMethod
    method()
  File "G:\FreeCAD\FreeCAD-build\Mod\Material\TestMaterialsApp.py", line 60, in testCalculiXSteel
    self.assertTrue(steel.hasAppearanceModel('f006c7e4-35b7-43d5-bbf9-c5d572309e6e')) # BasicRendering - inherited from Steel.FCMat
  File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 765, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true

19:19:47  
----------------------------------------------------------------------
Ran 5 tests in 0.0935s

19:19:47  FAILED (failures=1)
```
This will be tough to step through as currently structured because the only tests in Materials are in Python, and can't be run outside of FreeCAD. You might consider adding some C++ tests to see if you can reproduce the problem there.

**Comment by davesrocketshop** [comment]
> I'm trying to get these tests running on my system, and it's giving me some difficulty: this may or may not be related to the problem you are seeing (though I can reproduce that locally in a debug build, so I will continue to poke at it). Some observations:
> 

There are some changes that aren't in this branch. I was trying to correct them in a separate branch before merging, but I'll merge them shortly so you can test as well. Primarily it handles issues of '\\' and '/' in file paths.

I haven't been doing debug builds. I'll try that.

>     * yaml-cpp is being specified incorrectly in the cmake files, it isn't picking up the debug library when it should be, it always tries to use the release version. This might be a LibPack-specific problem.
> 
>     * When running TestMaterialsApp for the first time, I get the following output:
> 
> 
> ```
> 18:35:26  Empty filename passed to function
> 18:35:26  Empty filename passed to function
> 18:35:26  YAML parsing error: 'G:/FreeCAD/chennes/FreeCAD-chennes/src/Mod/Material/FluidMaterial/Air.FCMat'
> ```
> 
> I have no idea where that path is coming from, it's non-existent, and actually doesn't bear much resemblance to a real path on my system, and certainly no the installation path of this build.

The path is obtained in MaterialLoader::getMaterialLibraries(). The code is 

         QString resourceDir = QString::fromStdString(App::Application::getResourceDir()
                                                     + "/Mod/Material/Resources/Materials");

Yours is missing the "Resources" piece, so I'm not sure how or where that is happening.
> 
>     * When running tests for the first time, it takes a very long time to get started: maybe ten minutes? It is spending all that time in this loop:
> 
> 
> ```
> void ModelLoader::loadLibrary(const ModelLibrary& library)
> {
>     if (_modelEntryMap == nullptr) {
>         _modelEntryMap = new std::map<QString, ModelEntry*>();
>     }
> 
>     QDirIterator it(library.getDirectory(), QDirIterator::Subdirectories);
>     while (it.hasNext()) {
>         auto pathname = it.next();
>         QFileInfo file(pathname);
>         if (file.isFile()) {
> ```
> 
>     * At the end of the processing, I get an "Invalid string position" exception
> 
>     * If I run the tests a second time without restarting FreeCAD it runs much faster, but I get your failure from above:
> 

When run for the first time it loads and parses all of the material files. A delay is expected, but not 10 minutes. I have made no attempt to optimize this code yet choosing to make it stable first.

> 
> ```
> 19:19:47  FAILED: testCalculiXSteel (TestMaterialsApp.MaterialTestCases)
> Traceback (most recent call last):
>   File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 60, in testPartExecutor
>     yield
>   File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 676, in run
>     self._callTestMethod(testMethod)
>   File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 633, in _callTestMethod
>     method()
>   File "G:\FreeCAD\FreeCAD-build\Mod\Material\TestMaterialsApp.py", line 60, in testCalculiXSteel
>     self.assertTrue(steel.hasAppearanceModel('f006c7e4-35b7-43d5-bbf9-c5d572309e6e')) # BasicRendering - inherited from Steel.FCMat
>   File "G:\FreeCAD\FreeCAD-build\bin\lib\unittest\case.py", line 765, in assertTrue
>     raise self.failureException(msg)
> AssertionError: False is not true
> 
> 19:19:47  
> ----------------------------------------------------------------------
> Ran 5 tests in 0.0935s
> 
> 19:19:47  FAILED (failures=1)
> ```
> 
> This will be tough to step through as currently structured because the only tests in Materials are in Python, and can't be run outside of FreeCAD. You might consider adding some C++ tests to see if you can reproduce the problem there.

Are there workbenches using C++ tests? I'll look to find any but the only tests I see are Python based.

I've tried adding C++ debugging output using both the Base::Console() methods and fprintf(stdout, ...) methods, but all output is stripped from the log files so I see nothing. I'm flying blind here. When I add the current code it will have that debugging code in there which should help.

**Comment by chennes** [comment]
> The path is obtained in MaterialLoader::getMaterialLibraries(). The code is

I think that code should be using `getUserAppDataDir()`.

> Are there workbenches using C++ tests? I'll look to find any but the only tests I see are Python based.

Yes -- we use gtest, have a look in the "tests" top-level directory. I can help you get things set up if you want a hand.

**Comment by davesrocketshop** [comment]
> > The path is obtained in MaterialLoader::getMaterialLibraries(). The code is
> 
> I think that code should be using `getUserAppDataDir()`.

No, the intent here is to use the system directories from the installation. `getUserAppDataDir()` is used later in the same function for loading the user defined materials.

> 
> > Are there workbenches using C++ tests? I'll look to find any but the only tests I see are Python based.
> 
> Yes -- we use gtest, have a look in the "tests" top-level directory. I can help you get things set up if you want a hand.

This should be very handy. I'll check it out. Thanks!

**Comment by chennes** [comment]
> No, the intent here is to use the system directories from the installation. `getUserAppDataDir()` is used later in the same function for loading the user defined materials.

Got it -- it turns out that at least when running in the tests, `getMaterialLibraries` is not being called (at least, not before `loadLibrary`).

**Comment by chennes** [comment]
Here's the backtrace of the C++ calls, if it's any use to you (the Python is opaque):
```
>	Material_d.pyd!Materials::ModelLoader::loadLibrary(const Materials::ModelLibrary & library) Line 315	C++
 	Material_d.pyd!Materials::ModelLoader::loadLibraries() Line 348	C++
 	Material_d.pyd!Materials::ModelLoader::ModelLoader(std::map<QString,Materials::Model *,std::less<QString>,std::allocator<std::pair<QString const ,Materials::Model *>>> * modelMap, std::list<Materials::ModelLibrary *,std::allocator<Materials::ModelLibrary *>> * libraryList) Line 63	C++
 	Material_d.pyd!Materials::ModelManager::ModelManager() Line 53	C++
 	Material_d.pyd!Materials::ModelManager::getManager() Line 80	C++
```

**Comment by davesrocketshop** [comment]
I've been able to isolate the line of code that's crashing, but I have no idea why or what to do about it. It is this line from the `ModelLoader::getModelLibraries()` function:

         auto param = App::GetApplication().GetParameterGroupByPath(
            "User parameter:BaseApp/Preferences/Mod/Material/Resources");

 I've put it in a C++ unit test, and sure enough it segfaults every time, but only on Linux and not in my local builds.

I need to verify some changes but I'll push the unit test code shortly.

**Comment by davesrocketshop** [comment]
> I've been able to isolate the line of code that's crashing, but I have no idea why or what to do about it. It is this line from the `ModelLoader::getModelLibraries()` function:
> 
> ```
>      auto param = App::GetApplication().GetParameterGroupByPath(
>         "User parameter:BaseApp/Preferences/Mod/Material/Resources");
> ```

Turns out this was caused by not having an Application object initialized in the test framework. That will be in the next merge.

The other errors you were seeing (Empty file name, YAML errors, etc) were caused by the lack of a `user.cfg` file in a fresh install. A few simple checks fixed that.

**Comment by davesrocketshop** [comment]
It took a long time to find the problem and I still don't have a fix. Getting debug information out of CI was challenging to say the least. Adding C++ tests didn't help (at least not for this problem) because they passed fine while the Python was still failing.

The difference is in the paths used to access the models and materials files. For C++ tests, the following paths are used:

```
  Resource dir '/home/runner/work/FreeCAD/FreeCAD/build/share/'
  Application dir '/home/runner/work/FreeCAD/FreeCAD/build/'
  Library dir '/home/runner/work/FreeCAD/FreeCAD/build/lib'

```
For the Python tests, these paths are used:

```
  Resource dir /usr/local/share/
  Application dir /usr/local/
  Library dir /usr/local/lib

```
The resource files don't exist in the Resource folder, and I have no idea where they are actually placed. They're certainly not where they should be. I suspect something changed in the build system as this problem manifested after updating the branch.

Since the problem only occurs in the CI system and not in my builds, this has been a tedious process. It involves adding print statements, often in a convoluted way, waiting 10-15 minutes or more to see the CI build results, edit, repeat. If you make a small error, you have to endure another cycle. I've lost a week on this already. Le sigh.

The saga continues...

**Comment by 0penBrain** [comment]
> The difference is in the paths used to access the models and materials files. For C++ tests, the following paths are used:
> 
> ```
>   Resource dir '/home/runner/work/FreeCAD/FreeCAD/build/share/'
>   Application dir '/home/runner/work/FreeCAD/FreeCAD/build/'
>   Library dir '/home/runner/work/FreeCAD/FreeCAD/build/lib'
> ```
> 
> For the Python tests, these paths are used:
> 
> ```
>   Resource dir /usr/local/share/
>   Application dir /usr/local/
>   Library dir /usr/local/lib
> ```
> 
> The resource files don't exist in the Resource folder

The difference is because C++ tests run on the build directory, while Python tests run on the install.
If you miss some resources during Python tests, probably you have a missing or faulty INSTALL Cmake rule.

**Comment by davesrocketshop** [comment]
> > The difference is in the paths used to access the models and materials files. For C++ tests, the following paths are used:
> > ```
> >   Resource dir '/home/runner/work/FreeCAD/FreeCAD/build/share/'
> >   Application dir '/home/runner/work/FreeCAD/FreeCAD/build/'
> >   Library dir '/home/runner/work/FreeCAD/FreeCAD/build/lib'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For the Python tests, these paths are used:
> > ```
> >   Resource dir /usr/local/share/
> >   Application dir /usr/local/
> >   Library dir /usr/local/lib
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The resource files don't exist in the Resource folder
> 
> The difference is because C++ tests run on the build directory, while Python tests run on the install. If you miss some resources during Python tests, probably you have a missing or faulty INSTALL Cmake rule.

I suspect you're right. The install log certainly supports this. Investigating now...

**Comment by davesrocketshop** [comment]
Success!

Now to clean up all this debugging code...

**Comment by luzpaz** [comment]
Couldn't the debugging code be kept and just conditional if/when devs want to troubleshoot ?

**Comment by davesrocketshop** [comment]
> Couldn't the debugging code be kept and just conditional if/when devs want to troubleshoot ?

It's not all gone. But there was a lot of useless stuff in there. Not of it is difficult to add back in if required.

**Comment by 0penBrain** [comment]
Would be good to clean commit history. 😉

**Comment by davesrocketshop** [comment]
> Would be good to clean commit history. 😉

This is not something I've done before and therefore something I should learn. Any references? Messing with git timelines makes me nervous

**Comment by luzpaz** [comment]
I've used visual git tools like [gitkraken](https://www.gitkraken.com/) to work on cleaning git history before but even then it wasn't a massive PR like this one. I'm not sure how the pros do it with squashing many commits (some non-contiguous) in different parts of the code.

**Comment by 0penBrain** [comment]
> This is not something I've done before and therefore something I should learn. Any references? Messing with git timelines makes me nervous

Reading about `git squash` is generally a good idea.
You can checkout your branch to another one as a backup before testing. 😉

**Comment by chennes** [comment]
I'm happy to help with git squashing/merging -- I do it on the command line via an interactive rebase. I like to take small steps: rearrange commits into groups, make sure that works cleanly, then individually squash the groups, that sort of thing. As @0penBrain suggests, if you make a branch off of your work and do your changes there, you can non-destructively test.

**Comment by davesrocketshop** [comment]
> I'm happy to help with git squashing/merging -- I do it on the command line via an interactive rebase. I like to take small steps: rearrange commits into groups, make sure that works cleanly, then individually squash the groups, that sort of thing. As @0penBrain suggests, if you make a branch off of your work and do your changes there, you can non-destructively test.

Yep I think I need some help. I tried using online instructions and I'm pretty sure its all messed up now.

**Comment by chennes** [comment]
Did you make a branch before these changes? My local copy is a few days old.

**Comment by davesrocketshop** [comment]
> Did you make a branch before these changes? My local copy is a few days old.

I have a clean copy. Same branch but in a separate repository on a different machine.

**Comment by chennes** [comment]
OK, good. It looks like on this branch you've lost some commits -- were you working on the command line, or using a GUI of some kind?

**Comment by davesrocketshop** [comment]
Generally speaking I have no problems with CLI, but my lack of understanding of git's advanced features means multiple searches for relatively simple operations. So I tried using VSCode. I'm open to using either, but working through it with the CLI might be a good start.

Just sent you an email so we can take this offline..

**Comment by davesrocketshop** [comment]
Doing a squash on this branch proved problematic, as it's been active for a long period of time with a lot of cross merges. I've created a new "clean" branch that resolves those problems.

My question (knowing the probable answer) is can this PR be associated with the new branch, or should I open a new PR based on that branch and close this one?

**Comment by chennes** [comment]
I don't know of a way to re-associate a PR with a different branch -- probably best to just close this and make a new one. Just provide a backlink so future code archaeologists can figure out what's going on.

**Comment by davesrocketshop** [comment]
> I don't know of a way to re-associate a PR with a different branch -- probably best to just close this and make a new one. Just provide a backlink so future code archaeologists can figure out what's going on.

The answer I expected. On it...

**Comment by davesrocketshop** [comment]
Closed in favour of https://github.com/FreeCAD/FreeCAD/pull/10690

**Pull request closed without merging**